### ISSUE TITLE
module declaration problem

### ISSUE THREAD

**Issue by Zummer** [vytenisu/npm-dts#15]
Hi,
there is a problem

in version npm-dts@1.3.6

```
declare module '@xx/my-ui' {
  import main = require('@xx/my-ui/index');
  export = main;
}
```
and 

```
declare module'@xx/my-ui/index' {
   ... some exports
}
```

in new version npm-dts@1.3.9 add .ts to require path of main module

```
declare module '@xx/my-ui' {
  import main = require('@xx/my-ui/index**.ts**');
  export = main;
}
```

but module declaration is stil 
```
declare module'@xx/my-ui/index' {
   ... some exports
}
```

and I got an error when I try to import 

```
import {some} from '@xx/my-ui';
```

what can I do to resolve this problem?

**Comment by vytenisu** [comment]
Hi, I am not yet confident that I fully understand the context - might need a few additional details.

If I understand correctly - problem lies in the fact that ".ts" is added to one declaration file and not the other. Is that correct?

Where are "**" symbols coming from - are those also added by npm-dts?

Is @xx a package namespace or some alias?

May I ask to share source code which causes bad declaration files to be generated? If you do not want to share full source code, a simplified example code would help a lot. It would be best for me to reproduce the issue before attempting to fix it.

**Comment by Zummer** [comment]
1. problem lies in the fact that ".ts" is added to require path id  declaration.  that is correct
2. "**" not added by npm-dts, its added by github editor <ctrl+b>
3. @xx/my-ui package name

earlier I use 
import {some} from '@xx/my-ui';

but now I got error in project wich using that package
module "@xx/my-ui"
This module can only be referenced with ECMAScript imports/exports by turning on the 'allowSyntheticDefaultImports' flag and referencing its default export.ts(2497)

p.s. webpack generate index.js files, not index.ts

**Comment by Zummer** [comment]
my index.d.ts file: with 1.3.9

```
declare module'@xx/my-ui/index' {
 export { Component } from '@xx/my-ui/Components/Component';
  export { AnotherComponent } from '@xx/my-ui/Components/AnotherComponent';
}

declare module '@xx/my-ui' {
  import main = require('@xx/my-ui/index.ts');
  export = main;
}
```

why added .ts extension?

**Comment by vytenisu** [comment]
Thank you for additional details. I'll have a look sometime soon to understand what is going on and why. There has been a similar report here: https://github.com/vytenisu/npm-dts/issues/14. However, after fixing what seemed to be a plausible problem - author of the issue did not come back to confirm.

In the meantime please continue using older version which suits your needs. I will update this ticket once I have something.

**Comment by vytenisu** [comment]
Hello, got it. Sry I was not able to check the issue just yet. Will update here once I have something.

**Comment by WhiteTurbine** [comment]
Okay, I have investigated code, setup a separate project and was not able to reproduce the issue. This is the output I got:

index.d.ts

```
declare module '@smth/test/index' {
  export { TEST } from '@smth/test/lib/testing';

}
declare module '@smth/test/lib/testing' {
  export const TEST = "TEST";

}
declare module '@smth/test' {
  import main = require('@smth/test/index');
  export = main;
}
```

index.ts

`export {TEST} from './lib/testing'`

lib/testing.ts

`export const TEST = 'TEST'`

After checking code in detail I tried to understand under what circumstances npm-dts may leave ".ts" in the main import. It can happen if root address of the project is setup incorrectly. Script combines root address with main file (usually "index.ts") and gets a path to a file which does not exist. When path is invalid, script does not modify it to create a proper module name without extension and ".ts" gets left over.

In "npm-dts" by default ROOT address of the project is CWD. If script is launched from another directory, root may end up being wrong. However, you do have an option to set it up manually by passing "-r" modifier and a hard-coded path.

Same with "npm-dts-webpack-plugin". If your project has a more complicated structure, you may need to pass a configuration option "root" when you are initializing the plugin.

Root is the path to where your project's "package.json" is located. root + entry should be a valid path.

Please let me know if this helps. If not, please create some simple (as simple as possible) demo project in github where I could reproduce the issue. Maybe there is something wrong after all - I just could not see it yet.

**Comment by martinmckenna** [comment]
@WhiteTurbine here's a repro case:

https://github.com/martinmckenna/base-ts-module/tree/test

Running `yarn && yarn build` on that `test` branch there should give you a root level type declaration file with the issue described. My project does allow for absolute imports, but in my `index.ts` file, I'm using relative imports, so I don't think that's the issue.

Then if you want to checkout `master` which is on v1.2.1, you'll see things working correctly

**Comment by WhiteTurbine** [comment]
@martinmckenna , thank you for example - this is the best way to diagnose problems.

Basically what happens is **npm-dts** searches for "index.TS" in the root directory of your project and does not find it (index.js does not count). As a result it adds (leaves untouched) extension for "index.ts" in the main declaration.

To have nice paths you have alias for "src" and all your modules are named like 'my-package/something/inside/src". Aliases are not supported by **npm-dts** at the moment. Which makes me think it might be an important addition for the next version.

Possible solutions:
1. Since you are bundling module into 1 file, maybe you do not care that much whether there is "src" in path or not. After all, if your module is supposed to be used from a single entry point - end consumer does not see those paths. If having "src" in path is not a big issue, you can remove alias and use plugin in a following way:
```
  plugins: [
    new NpmDtsPlugin({
      logLevel: 'warn',
      entry: 'src/index.ts'
    })
  ]
```

2. If you do not like the idea of having "src" in paths, until aliases are supported there is 1 possible hack. Try creating empty "index.ts" file next to "index.js" in your root. This should cause script to generate proper main module declaration. But remember, this is a hack - not some supported solution :)

My takeaways from your example are:
- Would be great to support aliases so that main file would be searched for based on correct location.
- Need better documentation explaining that "entry" may be a path relative to root - not just a file name.
- Need a good warning message for when main file is not found where it is expected to be found.

Hope this helps. I will try to tackle above points when able. Leaving this issue open till then.

**Comment by martinmckenna** [comment]
that seems strange. i would assume most people have their `index.ts` file inside a `src` directory or something similar 🤔

**Comment by WhiteTurbine** [comment]
You can have it in src but without alias and with entry option for now as
per 1st fix suggestion.

**Comment by vytenisu** [comment]
After some investigation I have removed file presence check on entry file and forced removal of extension.

Fix version: **1.3.10**

I am closing this ticket for now. Please feel free to re-open should issue persist.
